### ISSUE TITLE
Improve map rendering speed (expected sub-second)

### ISSUE THREAD

**Issue by ioogithub** [osmandapp/OsmAnd#8370]
Device: Motorola G7
OS: Android 9 (LineageOS nightly)
Version: 3.3.5 (latest F-Droid version)

When I interact with the map it is very slow to respond. Specifically if I zoom out, the middle tile remains while all of the other tiles around it are black for around 3-5 seconds then it draws the map all at once.

If I zoom in, everything is blurry for 3-7 seconds then it either renders all at once or the elements slowly trickle in.

I can't image this is a limitation of the phone. It is a midrange phone but a current market version. All other software installed acts normally including another map application which zooms smooth and normally.

If I scroll off the screen the map can't keep up either, I see black boxes at the edges, the elements are blurry.

I have tried downloading offline maps, tried changing maps, using online maps, disabled animations, removed all layers and plug-ins except the base map but nothing helps.

I would say it is almost unusable like this.

**Comment by dbadura** [comment]
I have the same problem with slow rendering.
Android: Android 9 AESP
CPU: Snapdragon 635
No GSF on the phone

Here is a short video
https://gfycat.com/insistentbrownfairybluebird

**Comment by brodock** [comment]
Same for Huawei P20 Pro.

**Comment by sonora** [comment]
What you see in the video posted above is the effect of producing a map view from vectorized maps, which is needed for navigation and making all features of the map and their appearance fully configurable. Versus a so called raster or tile map, where you merely cache predefined images and display them on demand. This takes much more storage, is rigid in appearance, but fast.

In OsmAnd you can test by activating the built in Online maps plugin, then under Configure map/Map source select a tile map source. Please note that initially tiles have to be downloaded, so you would see the speed gain only after that.

Regarding the original post: How drastic is the issue reported? Is it like shown in the video, which seems the current state of the art for mid range hardware (including sd cards), or is it a lot worse?

**Comment by dbadura** [comment]
I think that the issue is the same like on the video.

I also tested the map on snapdragon 632 with normal ROM and the rendering speed was slow.

**Comment by dead10ck** [comment]
I experience this as well, just as bad as the video posted above, on a OnePlus 7T, which is Snapdragon 855+, top of the line, so it's definitely the app. The performance of the renderer is quite bad.

**Comment by dead10ck** [comment]
It's also worth noting that online maps don't seem to make much of a difference: https://gfycat.com/diligentfeistyizuthrush

**Comment by dead10ck** [comment]
Is any work being planned on this? This is one of the most common complaints about OSMAnd, next to the router doing weird things.

**Comment by brodock** [comment]
it's worth noting the iPad version is extremely fast and basically instant, this is really android specific.

**Comment by artfulrobot** [comment]
Love the app, *thanks!* But also finding the vector rendering slow. While I understand the benefits of it, I don't understand why the result of the render is not cached. If I scroll left, then right again it seems to have to redraw from the vector again, yet nothing has changed - surely a short lived cache of the render results (5 mins? 30 mins?) would really help?

**Comment by dead10ck** [comment]
I really don't think vector vs raster has anything to do with it, it's likely just the implementation. Almost every other navigation app uses vector graphics, and none of them are slow.

**Comment by sirius-c** [comment]
Indeed the render speed on a Nokia 5 is on the edge of usability. Regular render times as shown in debug mode of 3-5 seconds makes it appear very laggy. - Latest OsmAnd version an Android 9.

Is there some improvement work going on about this? Can other useful debug information be provided?

**Comment by CliveTheSecond** [comment]
I have a recording from my Pixel 2 XL running GrapheneOS (Android 11).
Towards the end you can see the whole map is pure white except a strip at the bottom which fully renders before the rest of the map appears.

Another thing to note is it works perfectly/instantly on an iPhone 8, so I agree with what @dead10ck said about raster vs vector.

Recording: https://tube.tchncs.de/videos/watch/3b407008-d683-4418-b146-acc52a98b84f

**Comment by artfulrobot** [comment]
Yep, @CliveTheSecond details my problem accurately. That video is exactly my experience (I'm on Android 8.1). Please ignore my raster/vector comment.

Sometimes during actual road navigation it can't draw the map fast enough to show me where I'm going, so I've had to go back to Google Maps, and not use the navigation function, which seems a shame. Sounds promising that the IOS version is smoother; makes it sound fixable.

**Comment by kaimast** [comment]
I can observe this as well on a Pixel 4a with Android 11.

Maybe a short term fix could be to add an option to increase the "view range" that would render tiles that are outside the screen? I also agree that caching would help a lot. Modern phones have a decent amount of (V)RAM, so keeping the tiles in memory should not cause any major problems. 

I also suspect that OsmAnd might not use the full concurrency available on modern multi-core systems. Maybe increase the number of threads that render the map tiles?

**Comment by fbanning** [comment]
I'm experiencing the same problem.

Device: Sony Xperia XA2
OS: LineageOS 17.1 (Nightly 20201210), no Google services
OsmAnd version: 3.8.5 (latest F-Droid)
Map source: Offline

Rendering times are pretty slow (~4-5 seconds) and this gets way worse when scrolling around a bit.

> caching would help a lot

Maybe, but there's a case to be made that throwing resources at it would only conceal the actual problem. If it is how others have stated (iOS version rendering very fast) then I would argue looking at this first is a lot more promising.

**Comment by Cliveptr** [comment]
I'm also thinking this could very likely be the major reason behind the battery drain some people have observed, though it should be noted I haven't looked at those issues at all.

IMO caching should only be considered after the underlying issue is fixed as it would only cover it up, but even then I'm not sure how much caching would help.

I'd say this is probably the second biggest reason I still have Google Maps on my phone.

**Comment by grravity** [comment]
Can confirm, also while driving the map constantly zooms in and out I assume when it wants to show you more roads, but when doing so the animation is extremely choppy and all the roads and names seem to bug out for a second when animating out and in.

Also gray blocks everywhere when it animates in and out.

Manually dragging the map, or zooming, also creates these gray blocks.

I tried vectorized online maps as well, didn't fix.

LGv40

**Comment by tegutistgut** [comment]
Can confirm,
Samsung Galaxy A51, Android 10, OneUi 2.5, normal samsung rom, no root.
Newest version of OsmAnd from the Google play store. 

Zooming in causes a lag and it takes about 3-5 seconds. Big deal breaker. 

Google Maps loads it in 1-2 seconds and with no lag.



Please fix this issue.

Thanks

**Comment by sonora** [comment]
To e everyone reporting here, can you please mention if your observation is in line with the video posted above https://github.com/osmandapp/OsmAnd/issues/8370#issuecomment-593098794, or if the issue is worse?

I would say that that video displays rendering times of 2 seconds at most (unless you keep panning and dont let map display finish), but some reports here talk about 3-5 sec? Also, I read about 'grey boxes', but that does not seem to march with what the video shows?

Essentially I am wondering if maybe sone reports refer to using 'safe mode' being set to 'on', or refer to the GL app version being used and lacking compatibility....

**Comment by sirius-c** [comment]
I call the behaviour seen in the video as almost snappy compared to my case (Nokia 5), where I tend to think often, while it is rendering, "come on, come on... you can do it, OSM". Two seconds until the map is completed would be an aceptable time for me, but five seconds is unbearable long.

**Comment by tegutistgut** [comment]
> To e everyone reporting here, can you please mention if your observation is in line with the video posted above https://github.com/osmandapp/OsmAnd/issues/8370#issuecomment-593098794, or if the issue is worse?
> 
> I would say that that video displays rendering times of 2 seconds at most (unless you keep panning and dont let map display finish), but some reports here talk about 3-5 sec? Also, I read about 'grey boxes', but that does not seem to march with what the video shows?
> 
> Essentially I am wondering if maybe sone reports refer to using 'safe mode' being set to 'on', or refer to the GL app version being used and lacking compatibility....

Here is a video made from my galaxy a51.

https://thumbs.gfycat.com/MammothCostlyHammerheadshark-size_restricted.gif

**Comment by sonora** [comment]
Ok, thanks! But that video again looks like the '2-seconds case', @tegutistgut I understand you are reporting something considerably slower?

**Comment by dead10ck** [comment]
@sonora I'm curious why you're trying to differentiate between 2 and 5 seconds? Anything on the order of seconds is pretty extremely poor performance. All of the videos posted in this thread show that.

**Comment by vshcherb** [comment]
Numbers are visible if you enable Development Plugin -> Rendering info.
- Our reference is 500 ms and 800 ms for rendering - so it's considered to be OK
- 1000 ms - 1500 ms with some exceptions of very densed area of 2500 ms is OK only for SLOW devices (so this is a grey area)
- > +5000ms - is considered to be VERY SLOW and it will be addressed. 

So this is a quick cheatsheet to check expectations  and how these issues will be addressed.

**Comment by dead10ck** [comment]
@vshcherb I'm not sure what your base line is for a "slow device" but it is just as bad for me as it is in all of these videos, and I'm running a OnePlus 7T, which was top of the line last year, so it is certainly not a "slow device."

**Comment by vshcherb** [comment]
It's going offtopic a bit. New devices usually much faster with rendering cause memory is more responsive.

I support the idea of having new rendering engine and it's development should be finished first. In sense of the issue shown on the video, we could only investigate if there are too much details on the map (so to remove something) and measure vs our reference top Android devices that we have such as (Samsung S10, S20), as a low-end device we use cheap Nokia android devices.

**Comment by sirius-c** [comment]
On my Nokia 5, I am usually in the range of a few seconds for large overviews or for rather detailed views. But for an intermediate zoom in Germany, for example, I just got 2.6 sec searching and 6.4 secs rendering. Those times remain this large when moving around at this zoom scale.

**Comment by dead10ck** [comment]
@vshcherb sorry, I'm a bit confused with your wording.

Yes, I understand that faster devices are faster. What I'm saying is I'm running a new fast device and it still renders super slowly. So which device is kind of irrelevant, since everyone in this thread saying they're experiencing this problem are not running super slow devices.

Are you saying you don't want to prioritize improving performance? And you mentioned a new rendering engine: are you talking about a new item of work, or is that something ongoing, and is there a ticket to track it?

**Comment by sonora** [comment]
Speaking for my involvement in the above thread, and the "2 vs. 5 seconds" question: 

I was trying to determine if this is issue is reporting
- an **anomaly** e.g. only affecting certain devices, OS versions, or configurations, which may result from an individual misconfiguration or a bug undrer certain circimstances, and we could try to patch,
- or if you are reporting behavior and performance I see in line with a wide range of test devices I benchmark OsmAnd on, where improvents beyond this characteristic performance would require a change in conept (like display fewer map details) or an altogether new rendering engine and corresponding software engineering.

From all I gather now this cannot  be classified as a bug report, but is rather a general request to improve the map performanc as observed on a wide range of devices.

**Comment by grravity** [comment]
This doesn't seem to be an anomaly, definitely variety.
However I personally think it should be high priority, it can be extremely distracting and is one of the core features of the app.

**Comment by vshcherb** [comment]
This is ongoing effort and related issues are 
https://github.com/osmandapp/OsmAnd/issues/1972
https://github.com/osmandapp/OsmAnd/issues/7162

**Comment by tegutistgut** [comment]
Any updates? Or a ETA when it's finished?
Thx

**Comment by tegutistgut** [comment]
Any updates? Come on, this bug is obviously not "nice to have".

Why do other map apps (here we go, g-maps, or even qwant web map or any map from the browser) are waaayy faster than osmand?


Please fix this asap

**Comment by scaidermern** [comment]
What is the actual bottleneck here? Is it I/O for reading of map files? Is it RAM for keeping the map files in memory?Is it CPU for rendering? Will switching to GPU-based rendering/OpenGL help? Any research on this topic?

**Comment by tegutistgut** [comment]
@scaidermern I dont know. I have a galaxy A51, with 4 GB of ram, Exynos 9611, Android 11, OneUI 3.1 and GoogleMaps, HereWeGo Maps and even Qwant Maps from the Firefox Browser are running perfectly. So i dont think it has anything to do with my hardware or software.

I really think its the devs "fault" that osmand is super slow at rendering Maps.

**Comment by somini** [comment]
> I really think its the devs "fault" that osmand is super slow at rendering Maps.

How does this kind of comment help? Anyone really believe the OsmAnd devs are boycotting their own program?

> What is the actual bottleneck here? Is it I/O for reading of map files? Is it RAM for keeping the map files in memory? Is it CPU for rendering? Will switching to GPU-based rendering/OpenGL help? Should the map format be changed to accelerate rendering and routing? Any research on this topic?

This is the only technical question that needs answering.

**Comment by tegutistgut** [comment]
Hello? Any updates?

**Comment by PieturP** [comment]
@vshcherb  I'd say this shouldn't be labelled `Nice to Have`. It is rendering the map way too slow to be useful. Hence it really is a bug.

**Comment by dbadura** [comment]
I would say that this issue should be at least `high`.

Whether any work is being carried out on this task?

**Comment by seniorm0ment** [comment]
I agree, this is a *major* core part of the app and usage, and clearly an issue affecting many people on many devices including flagships. I hope this is worked on and set at higer priority.

**Comment by tegutistgut** [comment]
Yes i hope this too. The app is literally not usable with this bug. Otherwise it would be one of the best Map apps for mobile devices.

**Comment by brodock** [comment]
I've learned about a OSM map editor today that runs on Android (GPLv3) and
to my surprise does render much faster (I believe it relies on vectors):

https://github.com/streetcomplete/StreetComplete

I hope whatever they are doing there could be useful here

Em qua, 9 de jun de 2021 20:01, tegutistgut ***@***.***>
escreveu:

**Comment by dead10ck** [comment]
There are many apps that render OSM maps. This performance problem is very unique to OSMAnd.
